### PR TITLE
Remove failed downloaded files immediately

### PR DESCRIPTION
**Pull request by def-**
Especially skins were only deleted after they were requested again

### CODE REVIEW

File: src/engine/client/http.cpp [hunk 1]
```diff
@@ -287,6 +287,20 @@ bool CGetFile::BeforeCompletion()
...
+	{
+		char aPath[512];
+		if(m_StorageType == -2)
+			m_pStorage->GetBinaryPath(m_aDest, aPath, sizeof(aPath));
+		else
+			m_pStorage->GetCompletePath(m_StorageType, m_aDest, aPath, sizeof(aPath));
+
+		m_pStorage->RemoveFile(aPath, IStorage::TYPE_ABSOLUTE);
+	}
+}
```
Comment (heinrich5991) [inline src/engine/client/http.cpp:302]: Store the path in `class CGetFile` instead, it seems likely to get out of sync this way.

Comment (def-) [reply src/engine/client/http.cpp:302]: Done

**Comment by heinrich5991** [comment]
bors r+

Thanks.

**Comment by Jupeyy** [comment]
this isnt really thread safe @def- 
my main thread is faster than the http one and opens the fallback download map from the gameserver before the http thread called Oncompletion, so it deletes the file that was opened by the main thread, an thus, the gameserver download will never be finished
=> results in
screenshot_2020-09-24_00-43-53

**Comment by def-** [comment]
Now I'm confused. We remove the file only when the download failed. Why does it fail and then still get used?

**Comment by Jupeyy** [comment]
it tries ddnet http first, fails
sets to HTTP_ERROR, and sets the atomic before oncomplition is called(which will delete the file)
then the main thread sees that the atomic changed, starts opening the mapfile for gameserver download fallback.
then oncomplition is called, deletes the file, which was just opened by the main thread
=> file is gone

**Pull request merged**